core/src/main/scalajvm/sttp/client4/internal/httpclient/WebSocketImpl.scala

```diff
@@ -55,29 +55,58 @@ private[client4] class WebSocketImpl[F[_]](
           // making close sequentially idempotent
           if (wasOpen) fromCompletableFuture(ws.sendClose(statusCode, reasonText)) else ().unit
       }
-    }))
+    })
 
   override lazy val upgradeHeaders: Headers = Headers(Nil)
 
   override def isOpen(): F[Boolean] = monad.eval(_isOpen.get())
 
-  override implicit def monad: MonadError[F] = _monad
-
-  private def fromCompletableFuture(cf: CompletableFuture[JWebSocket]): F[Unit] =
-    _monad.async { cb =>
-      cf.whenComplete(new BiConsumer[JWebSocket, Throwable] {
-        override def accept(t: JWebSocket, error: Throwable): Unit =
-          if (error != null) {
-            cb(Left(error))
-          } else {
-            cb(Right(()))
-          }
-      })
-      Canceler { () =>
-        cf.cancel(true)
-        ()
-      }
+  protected[client4] def fromCompletableFuture(cf: CompletableFuture[JWebSocket]): F[Unit]
+
+}
+
+private[client4] object WebSocketImpl {
+  def sync[F[_]](
+      ws: JWebSocket,
+      queue: SimpleQueue[F, WebSocketEvent],
+      _isOpen: AtomicBoolean,
+      sequencer: Sequencer[F],
+      _monad: MonadError[F]
+  ): WebSocketImpl[F] =
+    new WebSocketImpl[F](ws, queue, _isOpen, sequencer) {
+      override implicit def monad: MonadError[F] = _monad
+
+      override protected[client4] def fromCompletableFuture(cf: CompletableFuture[JWebSocket]): F[Unit] =
```

Comment:
[minor] maybe we should simply make `fromCompletableFuture` a parameter (of type `CF => F[Unit]`), instead of a method?